store/postgres/src/copy.rs

```diff
@@ -268,14 +273,106 @@ impl CopyState {
     }
 }
 
-struct TableState {
-    dst_site: Arc<Site>,
+/// Track the desired size of a batch in such a way that doing the next
+/// batch gets close to TARGET_DURATION for the time it takes to copy one
+/// batch, but don't step up the size by more than 2x at once
+#[derive(Debug, Queryable)]
+pub(crate) struct AdaptiveBatchSize {
+    size: i64,
+}
+
+impl AdaptiveBatchSize {
+    pub fn new(table: &Table) -> Self {
+        let size = if table.columns.iter().any(|col| col.is_list()) {
+            INITIAL_BATCH_SIZE_LIST
+        } else {
+            INITIAL_BATCH_SIZE
+        };
+
+        Self { size }
+    }
+
+    // adjust batch size by trying to extrapolate in such a way that we
+    // get close to TARGET_DURATION for the time it takes to copy one
+    // batch, but don't step up batch_size by more than 2x at once
+    pub fn adapt(&mut self, duration: Duration) {
+        let new_batch_size =
+            self.size as f64 * TARGET_DURATION.as_millis() as f64 / duration.as_millis() as f64;
```

Comment:
I see this code was just moved, but to make sure we avoid division by zero:
```suggestion
            self.size as f64 * TARGET_DURATION.as_millis() as f64 / (duration.as_millis() as f64 + 1);
```

Reply:
Nice idea, added